driver-sync/src/main/com/mongodb/client/internal/ClientSessionBinding.java

```diff
@@ -32,21 +38,46 @@
  * This class is not part of the public API and may be removed or changed at any time.
  */
 public class ClientSessionBinding implements ReadWriteBinding {
-    private final ReadWriteBinding wrapped;
+    private ReadWriteBinding wrapped;
     private final ClientSession session;
     private final boolean ownsSession;
     private final ClientSessionContext sessionContext;
+    private final Cluster cluster;
+    private ReadPreference readPreference;
 
     public ClientSessionBinding(final ClientSession session, final boolean ownsSession, final ReadWriteBinding wrapped) {
-        this.wrapped = notNull("wrapped", wrapped);
-        this.ownsSession = ownsSession;
         this.session = notNull("session", session);
+        this.ownsSession = ownsSession;
         this.sessionContext = new SyncClientSessionContext(session);
+        this.cluster = wrapped instanceof ClusterBinding ? ((ClusterBinding) wrapped).getCluster() : null;
+        this.wrapped = notNull("wrapped", initShardedTxnWrapped(wrapped));
+    }
+
+    private ReadWriteBinding initShardedTxnWrapped(final ReadWriteBinding wrapped) {
+        if (isActiveShardedTxn()) {
+            this.readPreference = session.getTransactionOptions().getReadPreference();
+            setPinnedMongosAddress();
+            wrapped.release();
+            return new SingleServerBinding(cluster, session.getPinnedMongosAddress(), readPreference);
+        }
+        return wrapped;
+    }
+
+    private boolean isActiveShardedTxn() {
+        return session != null && session.hasActiveTransaction() && cluster != null
```

Comment:
No need to check for null on session or cluster (see above comment)

Reply:
I have removed the null checks.